File: engine/core/src/elements/analog_switch2.rs
//! Analog switch with two throws (AnalogSwitch2Elm.java), the SPDT sibling
//! of the analog switch. Posts 0 is the common terminal, 1 and 2 the two
//! throws, 3 the control, which draws no current. The control gates which
//! throw carries `r_on`: below `threshold` the switch is open, so throw 2
//! carries the on-resistance and throw 1 the off-resistance, and FLAG_INVERT
//! swaps the pair. FLAG_PULLDOWN is inherited from the SPST (AnalogSwitchElm.
//! java:27): with it set, both throws are tied to ground through `r_off` for
//! the whole run, so the unselected throw is a complete disconnect instead of
//! carrying the off-resistance (AnalogSwitch2Elm.java:100-117).

use crate::element::{Base, Element, SimCtx};
use crate::spec::ElementSpec;
use crate::stamp::{Stamper, GROUND};

const DEF_R_ON: f64 = 20.0;
const DEF_R_OFF: f64 = 1e10;
const DEF_THRESHOLD: f64 = 2.5;

const FLAG_INVERT: i64 = 1; // AnalogSwitchElm.java:26
const FLAG_PULLDOWN: i64 = 2; // AnalogSwitchElm.java:27

/// Control-driven SPDT. The throw pairing changes with the control voltage
/// every step, so the element is nonlinear even though each stamp is a plain
/// resistor (AnalogSwitchElm.java:142).
pub struct AnalogSwitch2 {
    base: Base,
    r_on: f64,
    r_off: f64,
    threshold: f64,
    invert: bool,
    pulldown: bool,
    /// Whether the lever is open this step: throw 2 carries `r_on` when true,
    /// throw 1 when false.
    open: bool,
}

impl AnalogSwitch2 {
    pub fn new(spec: &ElementSpec) -> Self {
        Self {
            base: Base::with_posts(4),
            r_on: spec.param("r_on", DEF_R_ON),
            r_off: spec.param("r_off", DEF_R_OFF),
            threshold: spec.param("threshold", DEF_THRESHOLD),
            invert: spec.flag(FLAG_INVERT),
            pulldown: spec.flag(FLAG_PULLDOWN),
            open: true,
        }
    }

    /// The throw stamped with `r_on`. At rest (control below threshold) the
    /// switch is open, so post 2 is the normally closed side
    /// (AnalogSwitch2Elm.java:105-118).
    fn selected_post(&self) -> usize {
        if self.open {
            2
        } else {
            1
        }
    }
}

impl Element for AnalogSwitch2 {
    fn kind(&self) -> &'static str {
        "analogSwitch2"
    }
    fn base(&self) -> &Base {
        &self.base
    }
    fn base_mut(&mut self) -> &mut Base {
        &mut self.base
    }
    fn post_count(&self) -> usize {
        4
    }
    fn nonlinear(&self) -> bool {
        true
    }
    /// The control post never couples to the switch body, and the common and
    /// both throws always connect for matrix topology purposes, even though
    /// the chosen stamp changes every step (getConnection, AnalogSwitch2Elm.java:
    /// 124-126).
    fn connects(&self, a: usize, b: usize) -> bool {
        a != 3 && b != 3
    }

    fn stamp(&mut self, _ctx: &SimCtx, s: &mut Stamper) {
        // The pulldowns are constant for the whole run, so they belong in the
        // snapshot pass, exactly as the SPST's do (AnalogSwitchElm.java:149-153,
        // AnalogSwitch2Elm.java:100-103).
        if self.pulldown {
            s.resistor(self.base.nodes[1], GROUND, self.r_off);
            s.resistor(self.base.nodes[2], GROUND, self.r_off);
        }
    }

    fn do_step(&mut self, _ctx: &SimCtx, s: &mut Stamper) {
        self.open = self.base.volts[3] < self.threshold;
        if self.invert {
            self.open = !self.open;
        }
        // One throw carries `r_on`, the other `r_off`, whichever way the lever
        // points. With FLAG_PULLDOWN the unselected throw is already tied to
        // ground by `stamp`, so it is left a complete disconnect here
        // (AnalogSwitch2Elm.java:109-117).
        let sel = self.selected_post();
        let unsel = if sel == 1 { 2 } else { 1 };
        s.resistor(self.base.nodes[0], self.base.nodes[sel], self.r_on);
        if !self.pulldown {
            s.resistor(self.base.nodes[0], self.base.nodes[unsel], self.r_off);
        }
    }

    fn calculate_current(&mut self, _ctx: &SimCtx) {
        self.base.current =
            (self.base.volts[0] - self.base.volts[self.selected_post()]) / self.r_on;
    }

    fn current_into_node(&self, post: usize) -> f64 {
        if post == 0 {
            -self.base.current
        } else if post == self.selected_post() {
            self.base.current
        } else {
            0.0
        }
    }

    /// The quantity a voltage scope on the element should plot: the drop
    /// across the conducting throw, matching the current formula.
    fn voltage_diff(&self) -> f64 {
        self.base.volts[0] - self.base.volts[self.selected_post()]
    }

    fn set_param(&mut self, name: &str, value: f64) -> bool {
        match name {
            "r_on" if value > 0.0 => self.r_on = value,
            "r_off" if value > 0.0 => self.r_off = value,
            "threshold" => self.threshold = value,
            _ => return false,
        }
        true
    }

    fn reset(&mut self) {
        self.base.reset();
        self.open = true;
    }
}

File: web/src/model/registry/elements/analogSwitch2.ts
/**
 * Analog switch with two throws (AnalogSwitch2Elm.java), the SPDT sibling of
 * the analog switch: posts 0 common, 1 and 2 the two throws, 3 the control.
 * The control gates which throw carries `r_on`, so the common is never left
 * open. FLAG_PULLDOWN is inherited from the SPST (AnalogSwitchElm.java:27):
 * with it set, both throws ride `r_off` pulldowns to ground for the whole
 * run. The lever points at the conducting throw (AnalogSwitch2Elm.java:105-118).
 */

import {
  calcLeads,
  canvasFont,
  currentDots,
  dsign,
  elementLength,
  endpoints,
  interp,
  lead,
  line,
  voltageColor,
} from '../../../render/draw';
import { OPEN_HS, readParams, writeParams, postsBox } from '../shared';
import type { CircuitElement, DrawContext, ElementDef, Point } from '../../types';

const FLAG_INVERT = 1;  // AnalogSwitchElm.java:26
const FLAG_PULLDOWN = 2;  // AnalogSwitchElm.java:27

interface AnalogSwitch2Geometry {
  p1: Point;
  p2: Point;
  lead1: Point;
  lead2: Point;
  posts: Point[];
  poles: Point[];
  ctlPoint: Point;
}

/** The throw terminals, their fan points and the control post, matching
 *  AnalogSwitch2Elm.setPoints (AnalogSwitch2Elm.java:32-44): throw 1 hangs at
 *  `+openhs` on the `dsign`-determined side and throw 2 at `-openhs`, both at
 *  the far end; the control sits at the body midpoint on the `+openhs` side. */
function analogSwitch2Geometry(e: CircuitElement): AnalogSwitch2Geometry {
  const [p1, p2] = endpoints(e);
  const [lead1, lead2] = calcLeads(e, 32);
  const openhs = dsign(p1, p2) * OPEN_HS;
  return {
    p1,
    p2,
    lead1,
    lead2,
    posts: [p1, interp(p1, p2, 1, openhs), interp(p1, p2, 1, -openhs)],
    poles: [interp(lead1, lead2, 1, openhs), interp(lead1, lead2, 1, -openhs)],
    ctlPoint: interp(lead1, lead2, 0.5, openhs),
  };
}

function analogSwitch2Posts(e: CircuitElement): Point[] {
  const { posts, ctlPoint } = analogSwitch2Geometry(e);
  return [posts[0], posts[1], posts[2], ctlPoint];
}

/** The lever's throw index from the live control voltage, mirroring doStep
 *  (AnalogSwitch2Elm.java:105-108): open (control below threshold) throws to
 *  post 2, FLAG_INVERT swaps it. */
function leverPosition(g: DrawContext, e: CircuitElement): number {
  let open = (g.voltages[3] ?? 0) < (e.params.threshold ?? 2.5);
  if ((e.flags & FLAG_INVERT) !== 0) open = !open;
  return open ? 1 : 0;
}

function drawAnalogSwitch2(g: DrawContext, e: CircuitElement): void {
  const { p1, p2, lead1, posts, poles } = analogSwitch2Geometry(e);
  const position = leverPosition(g, e);

  // The common lead and the two throw leads, then the lever to the conducting
  // throw (AnalogSwitch2Elm.java:49-66). The control terminal has no stub
  // upstream; the dead-end post keeps its junction dot.
  lead(g, p1, lead1, voltageColor(g, g.voltages[0]));
  lead(g, poles[0], posts[1], voltageColor(g, g.voltages[1]));
  lead(g, poles[1], posts[2], voltageColor(g, g.voltages[2]));
  // The lever points at the conducting throw but does not carry its voltage:
  // upstream strokes it lightGray (AnalogSwitch2Elm.java:63-65).
  line(g, lead1, poles[position], g.theme.lightGray);

  currentDots(g, p1, lead1, g.current);
  currentDots(g, poles[position], posts[position + 1], g.current);

  // Name the throws when the part is picked, like upstream's highlight labels
  // (AnalogSwitch2Elm.java:75-80): at rest post 2 is NC and post 1 NO, unless
  // FLAG_INVERT swaps them.
  if (g.selected || g.hovered) {
    const dn = Math.max(1, elementLength(e));
    const off = OPEN_HS + 10;
    const labels = [
      interp(p1, p2, 1 - 10 / dn, off),
      interp(p1, p2, 1 - 10 / dn, -off),
    ];
    const inverted = (e.flags & FLAG_INVERT) !== 0;
    g.ctx.fillStyle = g.theme.selection;
    g.ctx.font = canvasFont(11);
    g.ctx.textAlign = 'center';
    g.ctx.textBaseline = 'middle';
    g.ctx.fillText(inverted ? 'NC' : 'NO', labels[0].x, labels[0].y);
    g.ctx.fillText(inverted ? 'NO' : 'NC', labels[1].x, labels[1].y);
  }
}

export const ANALOG_SWITCH2_DEF: ElementDef = {
  kind: 'analogSwitch2',
  label: 'Analog switch (SPDT)',
  category: 'Active',
  dumpCode: '160',
  postCount: 4,
  posts: analogSwitch2Posts,
  noDiagonal: true,  // AnalogSwitchElm.java:42
  defaultFlags: FLAG_PULLDOWN,  // the fresh constructor sets it (AnalogSwitchElm.java:43)
  defaults: { r_on: 20, r_off: 1e10, threshold: 2.5 },
  parse: (t, e) => readParams(t, e, ['r_on', 'r_off', 'threshold']),
  dump: writeParams(['r_on', 'r_off', 'threshold']),
  fields: [
    { name: 'r_on', label: 'On resistance', unit: 'Ω' },
    { name: 'r_off', label: 'Off resistance', unit: 'Ω' },
    { name: 'threshold', label: 'Threshold', unit: 'V' },
    { name: 'nc', label: 'Normally closed', type: 'bool', flag: FLAG_INVERT },
    { name: 'pulldown', label: 'Pulldown resistor', type: 'bool', flag: FLAG_PULLDOWN },
  ],
  // The two throws hang at ±openhs off the far end and the lever points at
  // one of them, so the whole fan is a solid pick zone (AnalogSwitch2Elm.java:
  // 48).
  bodyRect: (e) => postsBox(e, OPEN_HS),
  draw: drawAnalogSwitch2,
};
